### ISSUE TITLE
Command for traversing multi-project haskell modules

### ISSUE THREAD

**Issue by jhance** [yi-editor/yi#832]
Basically, I want to be able to start `yi` from within the project, and then something like `:module Yi.Editor` should bring me to the module for `Yi.Editor`. This will be especially convenient with the new project structure because it would allow the developer to care less about what package a particular module is when navigating.

A few brainstorming ideas: We can easily determine all of the projects that the user is interacting with by using stack as a library. Adding stack as a dependency is uncool but it could work in a package that isn't installed by default, and then hooked into the configuration by the user. Otherwise we probably have to traverse downwards, parse `stack.yaml`, and then decide the project structure whereas stack can probably do pretty much everything for us. We would then parse the cabal files for the individual projects by depending on `Cabal`.

We can kill the dependency on stack and retain the dependency on cabal with the following simpler strategy: Look downwards until we find a `stack.yaml`, then look up one folder in each subdirectory for a `cabal` file and use any that exist. This sounds nice, but is not quite as good as the stack-based approach because supposedly the stack library can autodetect a project structure when no `stack.yaml` exists. The upside is that the reduced dependency lets us include this in `yi-haskell-mode`.

(Question: How do we go about integrating keymaps, say a vim ex command, that requires something defined in `yi-haskell-mode`, with the new split project structure)?

**Comment by noughtmare** [comment]
I think the best solution is to make a new package `yi-mode-stack` and start stack mode when haskell mode is started and a stack.yaml is found like a minor mode in emacs. I don't think this requires anything defined in `yi-haskell-mode`.

**Comment by noughtmare** [comment]
We can start by depending on stack and eventually develop a 'microstack' or 'stack-lite' which only has the features we need.

**Comment by noughtmare** [comment]
We could use [commercialhaskell/intero](https://github.com/commercialhaskell/intero) for this instead of stack.

**Comment by jhance** [comment]
I think there might be a misunderstanding about how modes currently work. Our modes aren't as powerful as a typical emacs mode, and in general if a mode isn't interacting with syntax it doesn't need to be a mode. We also are limited to one mode per buffer (ideally we need to make modes composable).

Something like this is better exposed IMO as functionality that you can just hook into your keymap (that way it is agnostic between emacs and vim).

**Comment by noughtmare** [comment]
I don't think this is possible without writing a custom tool that can discover multi-package projects.

**Comment by jhance** [comment]
That is why I'm saying we can use stack, which can do that.

**Comment by shaleh** [comment]
Not sure why the comment is crossed out. Intero would definitely be the way to go. I am super impressed with how well it works in Emacs.

**Comment by noughtmare** [comment]
@shaleh I agree that it would be nice to have intero support in Yi, but it is not really relevent to this issue, because AFAIK intero has no commands for traversing multi-project haskell modules ([the tooling file of intero](https://github.com/commercialhaskell/intero/blob/master/TOOLING.md) doesn't mention it).

**Comment by shaleh** [comment]
As long as the stack config is responsible for the module, Intero indexes it and you could M-. to jump to a function definition.

**Comment by noughtmare** [comment]
I've looked into it again. You're right that it is pretty straightforward to use intero for this. The only problems I see are:
* It takes long to load larger project (because it required loaded packages to be compiled)
* It is not extensible, we only can use the builtin commands
* It requires communication via strings...

But these problems are not deal-breaking.

**Comment by noughtmare** [comment]
Another issue with intero is that if I load the Yi project (196 modules) in intero it uses 4GB of RAM.

**Comment by shaleh** [comment]
That is what bugs are for :-)

I have been very impressed using it in my (much smaller) projects.

**Comment by noughtmare** [comment]
This is implemented with `interoModule` in c26cef328f7226d0efb128cb850d75c1d10d51a5.

**Comment by jhance** [comment]
I'm not sure if "XXX being implemented via intero support" and "XXX being implemented without needing to run an external memory-hungry process just to locate files" are the same, so even with yi-intero I'd like to leave this open.
